README.md
Enhanced Sample Neufeld ASN C Compiler
=======================================

Build Status
------------
[![Travis Build Status](https://travis-ci.org/esnacc/esnacc-ng.png)](https://travis-ci.org/esnacc/esnacc-ng) | [![Build status](https://ci.appveyor.com/api/projects/status/f63n5ka5sqv93530/branch/master?svg=true)](https://ci.appveyor.com/project/orgcandman/esnacc-ng/branch/master)

What is eSNACC
--------------
This package originally provided the C/yacc/lex source code for snacc,
an ASN.1 to C or C++ compiler.  Snacc is short for "Sample Neufeld Asn.1
to C Compiler" and ASN.1 stands for Abstract Syntax Notation One (CCITT
X.208/ ISO 8824).  See the COPYRIGHT file for copyright information.

eSNACC is an enhanced version of the Sample Neufeld ASN C Compiler. It 
supports BER, DER, and PER rulesets for C and C++ code generation. There is
experimental support for TCL, and Python.

eSNACC takes ASN.1 code files, parses them, and outputs appropriate header
and source files for inclusion in a larger project.

eSNACC is licensed under two different liceses. The compiler itself is a
GPLv2+ licensed software; the version maintained here will continue to be
GPLv2, unless a compelling reason to switch presents itself. The runtime is
a much less restrictive license - it is considered as Public Domain code. So
if you wish to contribute, be sure that you understand any runtime changes
you contribute MUST be public domain for inclusion into this distribution. Any
compiler changes MUST be GPL, at least v2.

Building eSNACC
---------------
The following represents a very quick start guide to building the esnacc
codebase. For more detailed instructions, see [INSTALL.md]

First, run the './boot.sh' script, to generate the appropriate configure
scripts.

Next, run './configure' passing in the appropriate options.

Finally, run 'make' to build.

'make install' will put the binaries and library data in the appropriate
places.

Note: release and debug versions use the same library names.  If you want to be a
      debug version you must make sure you are working with a clean source tree.  
      Same goes for building a release version after building a debug version.


Running eSNACC
--------------

Given an ASN.1 source file(s) esnacc can produce:

   1. C routines for BER encoding, decoding, printing and freeing.
   2. C++ routines for BER encoding, decoding, and printing.
   3. Python routines for BER encoding, decoding, and printing.
   4. A type table that can be used with C driver routines
      for BER encoding, decoding, printing and freeing.
   5. if you are really lucky, a core dump! :)

esnacc tries to produce reasonably efficient C or C++ routines and data
structures to support BER encoding and decoding values of the given
ASN.1 data structures.  Printing routines for the decoded values are
also generated.  For C only, hierarchical freeing routines are
generated (but are not recommended for efficiency reasons).

When using the C or C++ (not table) options, esnacc creates the
following C files for each ASN.1 source file (module):

        1. .h file for the C/C++ translation of the ASN.1 data
           structure and prototypes for the generated routines.

        2. .c/.cpp file for the C/C++ encode, decode, print and free
           routines.

When using the table option, snacc produces a type table file
(a BER encoding) that can later be loaded at runtime by the
table driven encoding and decoding routines.

When using the python option, esnacc produces a .py file that can be 
used for BER encoding and decoding (sometimes).

FUTURE ENCHANCEMENTS
--------------------

* Support for 1997 ASN.1 syntax
* Change ANY tables so that each ASN.1 module is responsible for removing the 
  ANYs they install in the table.

PER FUTURE ENHANCEMENTS
-----------------------

 * PER 'C' code generation
 * Recognition and handling of extensibility


Features of ASN.1 Supported
---------------------------

   - parses ASN.1 '90 (subtype notation etc.)
   - macro definitions do not generate syntax errors but are
     are not processed. (retained as string)
   - value notation is parsed and for OBJECT IDENTIFIERs, INTEGERs
      and BOOLEANS (any other value in { }'s is currently kept as a string)
   - handles multiple ASN.1 module compiling/linking (IMPORTS/EXPORTS)
   - some X.400 and SNMP macros are parsed
   - supports ";" separted type or value definitions in the ASN.1 source.
     This is useful when dealing with some macros that introduce parsing
     problems.  Does not require the ";"'s though.
   - ANY DEFINED BY types are handled using the SNMP OBJECT-TYPE
     macro to define the identifier to type mapping.


NEWS
Future
------

* c++-lib: File Descriptor streambuffer
* c++-lib: streambuf monitor
* c++-lib: IOManipulator based encoding/decoding
* c++-lib: ROSE support infrastructure
* documentation: Developer's guides
* esnacc-logo: A mascot (and CC-BY license) was added for esnacc
* py-lib: Introduce a first cut at a python back-end
* py-gen: Add python backend, which supports only basic explicit tagging, no
          extensions, and BER only.


1.80
----

* Restarting full eSnacc support with eSnacc NG
* Converted to modern automake
* All c++ libraries have been renamed to cxx
* Fixed minbuf to follow the GenBuf abstraction
* Fixed GenBuf prototype
* Moved GenBuf macros to static inline functions
* Added support for modern style CONTAINs/SIZEs
* Fixed default value generation code for the C++ backend.
* Added experimental AUTOMATIC tags support

=======================================================================

New in 1.7:
 * Enchancements to C++ runtime:
 * Support of constraints checking for BER/PER
 * Added Asn-Relative-Oid's
 * Updated Asn-Oid to be inherited from Asn-Relative-Oid
 * Added Extensibility to the set/sequence/choice syntax (BER encoding/decoding only, PER is not yet supported)
 * asn::list has been changed to std::list (many changes)
 * non-optional set/sequence/choice elements are no longer generated as pointers
 * PER encode/decode capability for both aligned and unaligned variants (see PER beta notes below)
 * Enchancements to C runtime:
   * Added useful types
   * Added Asn-Relative-Oid's
 * Enhancements to compiler
   * eSNACC no longer supports un-named types (2002 syntax update)
   * added -b compiler option to turn on/off PER encoding/decoding function generation in set/sequence/choice (Note -- calling PEnc/PDec will still work, but will not produce correct encoding unless -b is used!!)

New in EKMS PER Beta (Packed Encoding Rules):
 * Aligned and unaligned PER variants (C++ only)
 * Limited constraint checking and PER encoding for PER visible constraints
 * Supported
   * char Stringtypes
   * Integer
   * Octet String
   * Bit String
   * Sequence-of / Set-of (limited)
   * Currently Unsupported 
     * wide char stringtypes
     * extensibility in constraints

New in 1.6Beta:
 * Updated "C" library to automatically handle ANY load/unloads as buffers.
 * Added interpretation of ASN.1 integer constants as values in tag references for "C" and C++.
 * Added "--snacc namespace: " pre-processor feature for unique C++ ASn.1 module namespace references.
 * Updated SNACC document (in the ./SNACC/doc directory) to present DigitalNet updates/enhancements.
 * Updated c++-examples and c-examples to demontrate recent features.

New in 1.5:
 * Updated "C" library to automatically handle ANY load/unloads as buffers.
 * Added interpretation of ASN.1 integer constants as values in tag references for "C" and C++.
 * Added "--snacc namespace: " pre-processor feature for unique C++ ASn.1 module namespace references.
 * Updated SNACC document (in the ./SNACC/doc directory) to present DigitalNet updates/enhancements.
 * Updated c++-examples and c-examples to demontrate recent features.


New in 1.4:
 * rewrote makefiles to make build process easier and faster.
 * Enchancements to C++ runtime:
   * AsnInt changed to be so that it no longer inherits AsnOcts
   * AsnBits enchanced to construct BitStrings from binary strings.
   * Added AsnSetOf and AsnSeqOf templates.
   * Added Exception handling (see snaccexcept.h)
   * Moved BDecPdu to AsnType.  So every type as access to it now.  This
     was done to help reduce the number of symbols & methods the compiler
     generates.
   * Added useful types
 * Enchancements to C runtime:
   * Added useful types
 * Enhancements to compiler
   * Removed -u switch because useful types are now in the runtime library.
   * Added useful types as basic types.

New in 1.3b4:
 * Generic table decoding routines with callback mechanism in tbl library
 * New tool "berdecode", decoding any BER data, can also use table file
 * New Tcl/Tk tool "asnwish", reads table files, access to grammar, en-/decoding
 * Fix for tag values > 2^14
 * Fix for table-encoding tag values dividable by 128
 * Small fix in AsnBits::GetBit to return TRUE (i.e., 1) instead of #bit in byte

New in 1.3b3:
 * Added isPdu flag to tables
 * Added number range checks during parsing

New in 1.3b2:
 * Small fix of C value string generation wrt char values >127
 * Added (limited) size constraints, bitstring and enumeration names to tables

New in 1.3b1:
 * Made snacc.h more C++-readable (credits to Steve Walker).
 * Improved dependency generation for stupid makedepends.
 * Corrected PeekTag to peek into buffer only as far as necessary.
 * Added installable error handler.
 * Fixed small glitch in idl-code generator (credits to Markku Savela).
 * Finally fixed cut-and-paste error in SEQUENCE OF parsing.

New in 1.3a<n>:
 * Continued porting C, C++, tbl to 64 bit, any endianness: Hash typedef,
   PeekTag algorithm used by tables.
 * Fixed cut-and-paste error in SEQUENCE OF parsing.
 * Added missing initialisation of exponent in asn-real.[Cc].
 * Fixed tbl-free.c and some other table-related bugs.
 * Added casts and stuff for picky C++ compilers.
 * Incorporated all valid bug fixes from snacc-bugs.
 * Initial move of tcl stuff to tcl7.6/tk4.2/tree4.1.

main changes between versions 1.1 and 1.2:

* ports to linux and alpha, which should make snacc both endianess and
  64 bit clean.
* to complement the destructors, T::T (const T&) and
  T &T::operator = (const T &) have been added to override the defaults
  supplied by the compiler. reason: simple pointer duplication may lead
  to unreferenced objects and to objects referenced more than once (on
  which the destructors delete may choke).
* the files asn_useful.[hcC] and tbl.[ch] get regenerated automagically.
  since the compiler itself is compiled with it, a bootstapping version
  is included in the distribution.

config_win32.h


/******************************************************************************/
/** RWC; VDA; This version was built for the Windows 95/NT development, not  **/
/** generated                                                                **/
/******************************************************************************/


/* config.h.  Generated automatically by configure.  */
/* config.h.in.  Generated automatically from configure.in by autoheader.  */

/* Define to empty if the keyword does not work.  */
/* #undef const */

/* Define to `int' if <sys/types.h> doesn't define.  */
/* #undef gid_t */

/* Define if you don't have vprintf but do have _doprnt.  */
/* #undef HAVE_DOPRNT */

/* Define if you have the vprintf function.  */
#define HAVE_VPRINTF 1

/* Define if your C compiler doesn't accept -c and -o together.  */
/* #undef NO_MINUS_C_MINUS_O */

/* Define if you need to in order for stat and other things to work.  */
/* #undef _POSIX_SOURCE */

/* Define to `unsigned' if <sys/types.h> doesn't define.  */
/* #undef size_t */

/* Define if you have the ANSI C header files.  */
/* #undef STDC_HEADERS */
#define STDC_HEADERS 1

/* Define if you can safely include both <sys/time.h> and <time.h>.  */
// RWC; #define TIME_WITH_SYS_TIME 1

/* Define to `int' if <sys/types.h> doesn't define.  */
/* #undef uid_t */

/* Define if your processor stores words with the most significant
   byte first (like Motorola and SPARC, unlike Intel and VAX). 
#define WORDS_BIGENDIAN 1
*/

/* Define if the X Window System is missing or not being used.  */
/* #undef X_DISPLAY_MISSING */

/* Define if lex declares yytext as a char * by default, not a char[].  */
#define YYTEXT_POINTER 1

/*
 * file: acconfig.h
 *
 * $Header: /baseline/SNACC/config_win32.h,v 1.2 2003/12/17 19:05:02 gronej Exp $
 * $Log: config_win32.h,v $
 * Revision 1.2  2003/12/17 19:05:02  gronej
 * SNACC baseline merged with PER v1_7 tag
 *
 * Revision 1.1.2.1  2003/11/05 14:58:53  gronej
 * working PER code merged with esnacc_1_6
 *
 * Revision 1.1.1.1  2000/08/21 20:35:45  leonberp
 * First CVS Version of SNACC.
 *
 * Revision 1.7  1997/03/03 11:58:26  wan
 * Final pre-delivery stuff (I hope).
 *
 * Revision 1.6  1997/02/28 13:39:34  wan
 * Modifications collected for new version 1.3: Bug fixes, tk4.2.
 *
 * Revision 1.5  1997/02/15 20:01:37  rj
 * check whether the compiler supports volatile functions (and whether abort() is volatile).
 *
 * Revision 1.4  1995/02/20  11:16:57  rj
 * cpp switch HAVE_VARIABLE_SIZED_AUTOMATIC_ARRAYS added.
 *
 * Revision 1.3  1995/02/13  14:46:49  rj
 * settings for IEEE_REAL_FMT/IEEE_REAL_LIB moved from {c_lib,c++_lib}/inc/asn_config.h to acconfig.h.
 *
 * Revision 1.2  1994/10/08  04:38:56  rj
 * slot for autoconf Tcl detection added.
 *
 * Revision 1.1  1994/09/01  00:51:19  rj
 * first check-in (new file).
 *
 */

/*
 * define IEEE_REAL_FMT if your system/compiler uses the native ieee double
 * this should improve the performance of encoding reals.
 * If your system has the IEEE library routines (iszero, isinf etc)
 * then define IEEE_REAL_LIB.  If neither are defined then
 * frexp is used.  Performance is probaby best for IEEE_REAL_FMT.
 *
 *  #define IEEE_REAL_FMT
 *  #define IEEE_REAL_LIB
 */
/* use ANSI or K&R style C? */
#define __USE_ANSI_C__ 1

/* does the C++ compiler have the bool type built-in? */
#define BOOL_BUILTIN 1

/* does the C++ compiler allow variable sized automatic arryas? */
#define HAVE_VARIABLE_SIZED_AUTOMATIC_ARRAYS 1

/* do we have all the libs we need for the Tcl interface? */
#undef HAVE_TCL

/* does the compiler support volatile functions (and is abort() volatile?) */
/* #undef COMPILER_WITHOUT_VOLATILE_FUNCTIONS */
/** RWC; 11/18/97 */
#undef COMPILER_WITHOUT_VOLATILE_FUNCTIONS

/* The number of bytes in a double.  */
#define SIZEOF_DOUBLE 8

/* The number of bytes in a int.  */
#define SIZEOF_INT 4

/* The number of bytes in a long.  */
#define SIZEOF_LONG 4

/* The number of bytes in a short.  */
#define SIZEOF_SHORT 2

/* Define if you have the finite function.  */
#define HAVE_FINITE 1

/* Define if you have the isinf function.  */
#define HAVE_ISINF 1

/* Define if you have the memcmp function.  */
#define HAVE_MEMCMP 1

/* Define if you have the memcpy function.  */
#define HAVE_MEMCPY 1

/* Define if you have the memset function.  */
#define HAVE_MEMSET 1

/* Define if you have the <dirent.h> header file.  */
#define HAVE_DIRENT_H 1

/* Define if you have the <fcntl.h> header file.  */
#define HAVE_FCNTL_H 1

/* Define if you have the <malloc.h> header file.  */
#define HAVE_MALLOC_H 1

/* Define if you have the <memory.h> header file.  */
#define HAVE_MEMORY_H 1

/* Define if you have the <ndir.h> header file.  */
/* #undef HAVE_NDIR_H */

/* Define if you have the <string.h> header file.  */
#define HAVE_STRING_H 1

/* Define if you have the <sys/dir.h> header file.  */
/* #undef HAVE_SYS_DIR_H */

/* Define if you have the <sys/ndir.h> header file.  */
/* #undef HAVE_SYS_NDIR_H */

/* Define if you have the <unistd.h> header file.  */
// RWC; #define HAVE_UNISTD_H 1

/* Define if you have the m library (-lm).  */
#define HAVE_LIBM 1
